source/agora/consensus/EnrollmentManager.d

```diff
@@ -568,6 +569,120 @@ public class EnrollmentManager
         return true;
     }
 
+    /***************************************************************************
+
+        Check if a pre-image exists
+
+        Params:
+            enroll_key = The key for the enrollment in which the pre-image is
+                contained.
+            height = The block height of the preimage to check existence
+
+        Returns:
+            true if the pre-image exists
+
+    ***************************************************************************/
+
+    public bool hasPreimage (const ref Hash enroll_key, ulong height) @trusted
+    {
+        auto results = this.db.execute("SELECT preimage from validator_set" ~
+                    " WHERE key = ?", enroll_key.toString());
+        if (!results.empty && results.oneValue!(byte[]).length != 0)
```

Comment:
If the result is not empty but the array length is zero, doesn't that mean there is a bug?

I think this should be turned into an assert.

```D
auto value = results.oneValue!(byte[]);
assert(value.length != 0);
auto loaded_image = value.deseralizeFull!PreimageInfo;
```